rust/hook-worker/src/worker.rs

```diff
@@ -265,51 +269,49 @@ async fn process_batch<'a>(
-
-                                let hog_function_id = metadata
-                                    .get("hogFunctionId")
-                                    .and_then(|h| h.as_str())
-                                    .map(|h| h.to_string())
-                                    .unwrap_or_else(|| "?".to_string());
-
-                                error!("dropping message due to size limit, team_id: {}, hog_function_id: {}", team_id, hog_function_id);
-                            }
-                            Err((error, _)) => {
-                                // Return early to avoid committing the batch.
-                                return log_kafka_error_and_sleep("send", Some(error)).await;
-                            }
-                        };
-                    }
+            Ok(result) if hog_mode => {
+                if let Some(payload) = create_hoghook_kafka_payload(result, &mut metadata).await {
+                    match kafka_producer.send_result(FutureRecord {
+                        topic: cdp_function_callbacks_topic,
+                        payload: Some(&payload),
+                        partition: None,
+                        key: None::<&str>,
+                        timestamp: None,
+                        headers: None,
+                    }) {
+                        Ok(future) => kafka_ack_futures.push(future),
+                        Err((
+                            KafkaError::MessageProduction(RDKafkaErrorCode::MessageSizeTooLarge),
+                            _,
+                        )) => {
+                            // HACK: While under development, we are dropping messages that
+                            // are too large. This is temporary, as we expect the webhook
+                            // handler for Hog to change soon. In the meantime, nobody needs
+                            // to be alerted about this.
+                            let team_id = metadata
+                                .get("teamId")
+                                .and_then(|t| t.as_number())
+                                .map(|t| t.to_string())
+                                .unwrap_or_else(|| "?".to_string());
+
+                            let hog_function_id = metadata
+                                .get("hogFunctionId")
+                                .and_then(|h| h.as_str())
+                                .map(|h| h.to_string())
+                                .unwrap_or_else(|| "?".to_string());
+
+                            error!("dropping message due to size limit, team_id: {}, hog_function_id: {}", team_id, hog_function_id);
+                        }
+                        Err((error, _)) => {
+                            // Return early to avoid committing the batch.
+                            return log_kafka_error_and_sleep("send", Some(error)).await;
+                        }
+                    };
                 }
             }
+            Ok(_) => {
```

Comment:
Nit: Generally think it's nicer if all the hogmode stuff lives in it's own function, since it's fairly self contained - something like:
```rust
    let results = join_all(futures).await;

    if hog_mode {
        // System relevant - this means that our requests are at-least once, since if we do the request, it succeeds, and then our kafka is down, we'll
        // do the request again. This was already true on batch commit, but now it's true on kafka send as well. We could add a "returned" state to the
        // state machine that indicates "we made the request but haven't pushed it to kafka yet", but we need to decide that's something we care about first.
        if let Err(_) = push_hoghook_results_to_kafka(results, metadata_vec, kafka_producer, cdp_function_callbacks_topic).await {
            return;
        }
    }

    let _ = batch.commit().await.map_err(|e| {
        error!("error committing transactional batch: {}", e);
    });
}

async fn push_hoghook_results_to_kafka(results: Vec<Result<WebhookResult, WorkerError>>, metadata_vec: Vec<Value>, kafka_producer: FutureProducer<KafkaContext>, cdp_function_callbacks_topic: &str) -> Result<(), ()> {
    let mut kafka_ack_futures = Vec::new();
    for (result, mut metadata) in iter::zip(results, metadata_vec) {
        match result {
            Ok(result) => {
                if let Some(payload) = create_hoghook_kafka_payload(result, &mut metadata).await {
                    match kafka_producer.send_result(FutureRecord {
                        topic: cdp_function_callbacks_topic,
                        payload: Some(&payload),
                        partition: None,
                        key: None::<&str>,
                        timestamp: None,
                        headers: None,
                    }) {
                        Ok(future) => kafka_ack_futures.push(future),
                        Err((
                            KafkaError::MessageProduction(RDKafkaErrorCode::MessageSizeTooLarge),
                            _,
                        )) => {
                            // HACK: While under development, we are dropping messages that
                            // are too large. This is temporary, as we expect the webhook
                            // handler for Hog to change soon. In the meantime, nobody needs
                            // to be alerted about this.
                            let team_id = metadata
                                .get("teamId")
                                .and_then(|t| t.as_number())
                                .map(|t| t.to_string())
                                .unwrap_or_else(|| "?".to_string());

                            let hog_function_id = metadata
                                .get("hogFunctionId")
                                .and_then(|h| h.as_str())
                                .map(|h| h.to_string())
                                .unwrap_or_else(|| "?".to_string());

                            error!("dropping message due to size limit, team_id: {}, hog_function_id: {}", team_id, hog_function_id);
                        }
                        Err((error, _)) => {
                            // Return early to avoid committing the batch.
                            return Err(log_kafka_error_and_sleep("send", Some(error)).await);
                        }
                    };
                }
            }
            Err(e) => {
                error!("error processing webhook job: {}", e)
            }
        }
    }

    for result in join_all(kafka_ack_futures).await {
        match result {
            Ok(Ok(_)) => {}
            Ok(Err((error, _))) => {
                // Return early to avoid committing the batch.
                return Err(log_kafka_error_and_sleep("ack", Some(error)).await);
            }
            Err(Canceled) => {
                // Cancelled due to timeout while retrying
                // Return early to avoid committing the batch.
                return Err(log_kafka_error_and_sleep("timeout", None).await);
            }
        }
    }
    Ok(())
}
```

Reply:
Yeah, this was really ugly, great idea.